---
title: "Lab 6"
author: "Maria Bassem_20011141"
date: "2024-04-15"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part 1: Quality Control and linkage disequilibrium (LD) with PLINK (10 Points)

### Task 1.1: QC, LD, and PCA (10 Points)

#### Run the PLINK command that filters out variants below the minor allele frequency threshold, excludes samples with low genotyping rates, and removes SNPs not in Hardy-Weinberg equilibrium using the following thresholds (hwe: 0.00001, maf: 0.05, geno: 0.001). The output is called cleaned data.

`plink --bfile Qatari156_filtered_pruned --maf 0.05 --geno 0.001 --hwe 0.00001 --make-bed --out cleaned_data`

![](images/file_show_path=~_2Fimages_2Fclipboard-1333706329.png&id=33.png)

#### On the cleaned data, run the following command to perform linkage disequilibrium pruning. The output is called pruned_data

`plink --bfile cleaned_data --indep-pairwise 100 5 0.1 --out pruned_data`

**`--indep-pairwise 100 5 0.1`**: This flag instructs PLINK to perform LD [linkage disequilibrium] pruning. The three parameters control the pruning process:

-   **100**: Specifies the window size (number of variants) for sliding windows across the genome. Variants within this window are considered for LD calculation.

-   **5**: Sets the number of variants to shift the window at each step.

-   **0.1**: Sets the threshold for LD (Linkage Disequilibrium). Pairs of variants with an r\^2 value above this threshold are considered to be in LD and are candidates for removal.

![](images/file_show_path=~_2Fimages_2Fclipboard-3006556816.png&id=35.png)

#### Experiment with the thresholds 100, 5, and 0.1 and comment on the effect of increasing or decreasing the threshold.

[Window size = 200]{.underline}

`plink --bfile cleaned_data --indep-pairwise 200 5 0.1 --out pruned_data_window_200`\
By increasing the window size, PLINK considers a larger region for LD pruning. This leads to pruning more variants from chromosomes.

![](images/window_200.png)

[Step = 3]{.underline}

`plink --bfile cleaned_data --indep-pairwise 100 3 0.1 --out pruned_data_step_3`\
Reducing the step size allows for capturing more accurate LD (correlations) structures but could increase computational time as a drawback.

![](images/step_3.png)

[r\^2 value = 0]{.underline}

`plink --bfile cleaned_data --indep-pairwise 100 5 0 --out pruned_data_ld_0`

The r\^2 value equals 0, this means that we want to prune all the combinations of the SNPs (we keep the independent or correlated). In other words, PLINK will keep variants that are [not in LD with each other]{.underline}. In other words, variants with r\^2 value less than or equal to the 0 will be retained, while variants with r\^2 value above 0 will be pruned or removed.

![](images/ld_0.png)

[value = 0.5]{.underline}

`plink --bfile cleaned_data --indep-pairwise 100 5 0.5 --out pruned_data_ld_0.5`\
As the r\^2 value increases, this means that we remove the variants with high LD correlation, we leave the variants with r\^2 value \>= 0.5. It seams that our data is not highly correlated.

![](images/ld_0.5.png)

#### Run PCA on the pruned data using the PLINK and --pca flag

`plink --bfile cleaned_data --extract pruned_data.prune.out --pca --out pca_results`

Explanation:

-   `--bfile cleaned_data`: Specifies the original input dataset in PLINK binary format (.bed, .bim, .fam) named "cleaned_data".

-   `--extract pruned_data.prune.out`: Specifies the list of SNPs to keep, which is generated by the LD pruning step and saved in the file "pruned_data.prune.out".

-   `--out pca_results`: Specifies the name for the output files.

    ![](images/clipboard-3942212559.png)

## Part 2: Identify SNPs associated with population structure (40 Points)

### Task 2.1: Identify SNPs associated with genomic PCs using linear Regression Analysis (20 Points)

### Conduct linear regression analysis for all SNPs in the cleaned data (NOT the pruned data), with the model specified below

#### Perform the following steps:

-   Use Plink to recode your data to the 0, 1, 2 format with the --recode A option. This will generate a .raw file

`plink --bfile cleaned_data --recode A --out recoded_data`

The --recode A option in PLINK generates a text file in the "allele format", which includes an additive component (0/1/2) representing the number of alternate alleles present, without considering dominance. By default, the alleles are counted based on the A1 allele (minor allele). So, in summary, --recode A generates an allele file with an additive (0/1/2) component.

![](images/clipboard-2118653262.png)

[**Note:**]{.underline}

I cleaned the data further by changing the qc parameters so that the lm and glm can run on my device

The data is now 14,079 variants.

*New cleaned data:*

![](images/cleaned_data%2014,000.png)

*recoded data from the new cleaned data:*

![](images/recode%2014,000.png)

-   Read the .raw file into R (use the function read.table and specify header=T and sep=””)

    ```{r}
    # Set the file path
    file_path <- "D:\\Third Year Computer\\Term 2\\Bio\\Labs\\Lab 6\\mmc2\\recoded_data.raw"

    # Read the .raw file into R
    data <- read.table(file_path, header = TRUE, sep = "")

    head(data[, 1:6])
    ```

-   Isolate the columns that contain the SNP data, which usually follows the first six columns (use the select function from tidyverse). Keep the IDs to match those with the IDs of the PCs

```{r}

library(tidyverse)

snp_data <- data %>%
  select(1:2, 7:ncol(data))
```

-   Read the PCA eigenvalues and eigenvectors into R (use the function read.table)

```{r}
pca_eigenval <- read.table("D:\\Third Year Computer\\Term 2\\Bio\\Labs\\Lab 6\\mmc2\\pca_results.eigenval", header = FALSE, sep = "")


pca_eigenvec <- read.table("D:\\Third Year Computer\\Term 2\\Bio\\Labs\\Lab 6\\mmc2\\pca_results.eigenvec", header = FALSE, sep = "")
colnames(pca_eigenvec) <- c("Family ID", "Sample ID", paste0("PC", 1:(ncol(pca_eigenvec) - 2)))

head(pca_eigenval)
head(pca_eigenvec)

```

-   Create a nested loop to run linear regressions for the 3 PCs with all SNPs. Run the following for each PC with a single SNP, while correcting the model for the other two PCs.

```{r}
# initialize dataset
data_set <- cbind(pca_eigenvec[, c("PC1", "PC2", "PC3")], snp_data[,3:ncol(snp_data)])
head(data_set[, 1:4])
```

```{r}
# Initialize an empty data frame to store coefficients
sig_assoc_lm <- data.frame(pc = character(), rsid = character(), beta = numeric(), std_error = numeric(), p_value = numeric())

# Initialize a data frame to store p-values and SNP columns
pvalues_snp_df_lm <- data.frame(pc = character(), rsid = character(), p_value = numeric())

# Bonferroni correction = 0.05/#tests 
bonferroni_threshold <- 0.05 / (ncol(snp_data) - 2) # ((nrow(snp_data) - 2) * 3) 

for(pc in paste0("PC", 1:3)){
  for(snp_x in names(snp_data)[-(1:2)]){ 
    formula <- paste(pc, "~", snp_x, "+", paste(setdiff(c("PC1", "PC2", "PC3"), pc), collapse = " + "))
    lm_result <- lm(as.formula(formula), data = data_set)
    p_value <- summary(lm_result)$coefficients[2, "Pr(>|t|)"]
    # to remove the _T for example
    snp_removed <- gsub("_.*", "", snp_x)
    pvalues_snp_df_lm <- rbind(pvalues_snp_df_lm, data.frame(pc = pc, rsid = snp_removed, p_value = p_value))
    
    if(p_value < bonferroni_threshold){
      sig_assoc_lm <- rbind(sig_assoc_lm, data.frame(pc = pc, rsid = snp_removed, beta = summary(lm_result)$coefficients[2, "Estimate"], std_error = summary(lm_result)$coefficients[2, "Std. Error"], p_value = p_value))
      # cat(pc, ", SNP = ", snp_x, ", p_value = ", p_value, "\n")
    }
  }
}
```

```{r}
head(pvalues_snp_df_lm)

```

-   Determine all associations that pass a Bonferonni threshold of significance. What is this threshold?

This is a method to correct the significance threshold for multiple hypothesis testing.

`bonferroni_threshold <- 0.05 / ((ncol(snp_data) - 2) * 3)`

```{r}
top_10_pcs_sig_snps <- sig_assoc_lm %>%
  group_by(pc) %>%
  slice_min(order_by = p_value, n = 10)
head(top_10_pcs_sig_snps)
```

### Task 2.2: Identify SNPs that associate with the population subgroups (clusters) using logistic regression (20 Points)

Implement logistic regression models to cluster data using 3 highest variance PCA components assuming the number of clusters = 3 (i.e. k = 3). Perform the following steps:

```{r}
#install.packages("caret")
library(caret)
library(scatterplot3d)
```

-   Set your pseudorandom seed before running the clustering algorithm. (use set.seed)
-   Run the k-means clustering algorithm using the first 3 PCs. Plot the 3D scatterplot with different colors for each cluster. (use scatterplot3d)

```{r}
set.seed(1)  # Set seed for reproducibility
k <- 3  # Number of clusters
nstart <- 25  # Set nstart

cluster_result <- kmeans(pca_eigenvec[, c("PC1", "PC2", "PC3")], centers = k, nstart = nstart)

# Plot 3D scatterplot with cluster colors
scatterplot3d(pca_eigenvec$PC1, pca_eigenvec$PC2, pca_eigenvec$PC3, color = cluster_result$cluster, xlab = "PC1", ylab = "PC2", zlab = "PC3", main = "3D Scatterplot with Clusters")
```

-   Create a new dataframe of the 3 PCs and the one-hot encoding of cluster labels like the following illustration. (use dummyVars from the caret package to do so). You should have 6 columns in this dataframe.

```{r}
# factor() converts the vector into a factor variable. The function assigns a unique numeric value to each distinct level or category present in the vector, and these numeric values are used internally to represent the categories. It also shows the levels {1, 2, 3}
cluster_labels <- factor(cluster_result$cluster)
# dummyVars() --> separate each factor into its own column. 
cluster_dummy <- predict(dummyVars(~., data = data.frame(cluster = cluster_labels)), newdata = data.frame(cluster = cluster_labels))

cluster_df <- cbind(pca_eigenvec[, c("PC1", "PC2", "PC3")], cluster_dummy)
head(cluster_df)
```

-   Create a nested loop to run logistic regressions for each of the 3 clusters with ALL SNPs including PCs as covariates. Store the summary of each regression in a list for subsequent analysis and comparison.

```{r warning=FALSE}

# Initialize an empty data frame to store coefficients
sig_assoc_glm <- data.frame(cluster = character(), rsid = character(), beta = numeric(), std_error = numeric(), p_value = numeric())

pvalues_snp_df_glm <- data.frame(cluster = character(), rsid = character(), p_value = numeric())
# Bonferroni threshold
bonferroni_threshold <- 0.05 / (nrow(snp_data) - 2) # ((nrow(snp_data) - 2) * 3) 
# Initialize the data called cluster_data
cluster_data <- cbind(cluster_df, snp_data)

for(cl in 1:k){
  for(snp_column in names(snp_data)[-(1:2)]){
    formula <- paste(paste0("cluster.", cl), "~", snp_column)

    logistic_model <- glm(as.formula(formula), family = binomial, data = cluster_data)
    p_value <- summary(logistic_model)$coefficients[2, "Pr(>|z|)"]
    
    snp_removed <- gsub("_.*", "", snp_column)
    pvalues_snp_df_glm <- rbind(pvalues_snp_df_glm, data.frame(cluster = paste0("Cluster.", cl), rsid = snp_removed, p_value = p_value))
    
    if(p_value < bonferroni_threshold){
      sig_assoc_glm <- rbind(sig_assoc_glm, data.frame(cluster = cl, rsid = snp_removed, beta = summary(logistic_model)$coefficients[2, "Estimate"], std_error = summary(logistic_model)$coefficients[2, "Std. Error"], p_value = p_value))
      # cat("cluster.", cl, ", SNP = ", snp_column, ", p_value = ", p_value, "\n")
    }
  }
}
```

```{r}
head(pvalues_snp_df_glm)
```

-   Identify significant SNPs that are associated with each of the 3 clusters using a Bonferroni threshold

```{r}
top_10_clusters_sig_snps <- sig_assoc_glm %>%
  group_by(cluster) %>%
  slice_min(order_by = p_value, n = 10)
head(top_10_clusters_sig_snps)
```

### Task 3: Manhattan, Annotation, and Discussion (25 points)

#### Plot a Manhattan plot using the [qqman]{.underline} package in R for the results you got (you should use all association results you have for ALL SNPs, but you can select those that pass the threshold of 0.05 if your data is too huge for the memory). This will need preparing a file with SNP chromosome and position (get from the map file) and matching those with the p-values obtained from regression. Do this once for the results from Task 2.1 and Task 2.2 separately.

```{r}
library(tidyverse)

# Set the file path
map_file_path <- "D:\\Third Year Computer\\Term 2\\Bio\\Labs\\Lab 6\\mmc2\\recoded_data_ped_map.map"

# Read the .raw file into R
map_data <- read.table(map_file_path, header = FALSE)
colnames(map_data) <- c("CHR", "rsid", "GD", "BP")

# Select only the required columns
map_data <- map_data[, c("rsid", "CHR", "BP")]

# Merge p-values data frame from pcs with map data based on rsid
merged_data_lm <- merge(map_data, pvalues_snp_df_lm, by = "rsid", all.x = TRUE)
# Display the first few rows of the merged data
head(merged_data_lm)

# Merge p-values data frame from clusters with map data based on rsid
merged_data_glm <- merge(map_data, pvalues_snp_df_glm, by = "rsid", all.x = TRUE)
# Display the first few rows of the merged data
head(merged_data_glm)

```

Manhattan For PCs

```{r}
# install.packages("qqman")
library(qqman)

# Unique PCs
unique_pcs <- unique(merged_data_lm$pc)

# Plot Manhattan plot for each PC
for(pc in unique_pcs) {
  # Subset data for the current PC
  pc_data <- merged_data_lm[merged_data_lm$pc == pc, ]
  
  main_title <- paste("Manhattan Plot for", pc)
  # Plot the Manhattan plot
  manhattan(pc_data, chr="CHR", bp="BP", snp="rsid", p="p_value", col = c("blue", "orange"), main = main_title)
}
```

Manhattan For Clusters

```{r}
# Unique clusters
unique_clusters <- unique(merged_data_glm$cluster)
# Plot Manhattan plot for each PC
for(cl in unique_clusters) {
  # Subset data for the current PC
  cluster_data <- merged_data_glm[merged_data_glm$cluster == cl, ]
  
  main_title <- paste("Manhattan Plot for", cl)
  # Plot the Manhattan plot
  manhattan(cluster_data, chr="CHR", bp="BP", snp="rsid", p="p_value", col = c("blue", "orange"), main = main_title)
}
```

#### Use dbSNP from NCBI website to check the information on the 10 most significant SNPs, by doing the following steps for both results obtained from Task 2.1 and Task 2.2 separately:

Report a table of the 10 most significant SNPs that includes:

-   Chromosome (CHR) of the SNP

-   Position (BP) of the SNP

-   Statistics from regression results (beta coefficient, standard error and p-value)

-   From dbSNP, the gene that contains the SNP

-   From dbSNP, the allele frequency in different databases

```{r}
top_10_merged_sig_lm <- merge(top_10_pcs_sig_snps, map_data, by = "rsid", all.x = TRUE)

top_10_merged_sig_lm <- top_10_merged_sig_lm %>%
  arrange(pc)

# # Display the first few rows of the merged data
# head(top_10_merged_sig_lm)

## ---------------- separator ---------------- 

top_10_merged_sig_glm <- merge(top_10_clusters_sig_snps, map_data, by = "rsid", all.x = TRUE)

top_10_merged_sig_glm <- top_10_merged_sig_glm %>%
  arrange(cluster)
# 
# # Display the first few rows of the merged data
# head(top_10_merged_sig_glm)
```

PC1:

<table>
<thead>
<tr class="header">
<th><p>SNP</p></th>
<th><p>CHR</p></th>
<th><p>BP</p></th>
<th><p>Beta</p></th>
<th><p>P-Val</p></th>
<th><p>SE</p></th>
<th><p>Gene</p></th>
<th><p>Freq</p></th>
<th><p>Max Freq</p></th>
<th><p>Min Freq</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>rs12613530</p></td>
<td><p>2</p></td>
<td><p>31307461</p></td>
<td><p>0.12028175</p></td>
<td><p>9.005544e-19</p></td>
<td><p>0.011864217</p></td>
<td><p>CAPN14</p></td>
<td><p>T=0.058562/1846 <a href="https://www.ncbi.nlm.nih.gov/snp/rs12613530#frequency_tab">ALFA</a><br />
T=0.010491/47 Estonian<br />
T=0.014028/14 GoNL</p></td>
<td><p>0.058562</p></td>
<td><p>0.010491</p></td>
</tr>
<tr class="even">
<td><p>rs213037</p></td>
<td><p>1</p></td>
<td><p>21528471</p></td>
<td><p>0.09544822</p></td>
<td><p>1.729533e-16</p></td>
<td><p>0.010297216</p></td>
<td><p>ECE1</p></td>
<td><p>A=0.098461/20094 <a href="https://www.ncbi.nlm.nih.gov/snp/rs213037#frequency_tab">ALFA</a><br />
A=0.037037/2 PRJEB36033<br />
A=0.061667/37 NorthernSweden</p></td>
<td><p>0.098461</p></td>
<td><p>0.037037</p></td>
</tr>
<tr class="odd">
<td><p>rs235679</p></td>
<td><p>6</p></td>
<td><p>124304070</p></td>
<td><p>0.08930758</p></td>
<td><p>6.923165e-13</p></td>
<td><p>0.011766085</p></td>
<td><p>NKAIN2</p></td>
<td><p>C=0.012011/237 <a href="https://www.ncbi.nlm.nih.gov/snp/rs235679#frequency_tab">ALFA</a><br />
A=0.002396/7 KOREAN<br />
A=0.002725/46 TOMMO</p></td>
<td><p>0.012011</p></td>
<td><p>0.002396</p></td>
</tr>
<tr class="even">
<td><p>rs3181360</p></td>
<td><p>9</p></td>
<td><p>116731379</p></td>
<td><p>0.09159005</p></td>
<td><p>2.437532e-12</p></td>
<td><p>0.012342818</p></td>
<td><p>TNFSF8 </p></td>
<td><p>T=0.074084/24101 <a href="https://www.ncbi.nlm.nih.gov/snp/rs3181360#frequency_tab">ALFA</a><br />
T=0./0 PRJEB36033<br />
T=0.051402/11 Vietnamese</p></td>
<td><p>0.074084</p></td>
<td><p>0</p></td>
</tr>
<tr class="odd">
<td><p>rs359428</p></td>
<td><p>5</p></td>
<td><p>173190227</p></td>
<td><p>0.09359099</p></td>
<td><p>3.123404e-12</p></td>
<td><p>0.012008814</p></td>
<td><p>-</p></td>
<td><p>A=0.129774/3330 <a href="https://www.ncbi.nlm.nih.gov/snp/rs359428#frequency_tab">ALFA</a><br />
T=0./0 KOREAN<br />
A=0./0 Korea1K</p></td>
<td><p>0.129774</p></td>
<td><p>0</p></td>
</tr>
<tr class="even">
<td><p>rs3787397</p></td>
<td><p>20</p></td>
<td><p>52700084</p></td>
<td><p>0.08918383</p></td>
<td><p>3.174304e-12</p></td>
<td><p>0.00893237</p></td>
<td><p>DOK5 </p></td>
<td><p>A=0.33283/51972 <a href="https://www.ncbi.nlm.nih.gov/snp/rs3787397#frequency_tab">ALFA</a><br />
A=0.255591/160 Chileans<br />
G=0.261905/11 Siberian</p></td>
<td><p>0.33283</p></td>
<td><p>0.255591</p></td>
</tr>
<tr class="odd">
<td><p>rs4670713</p></td>
<td><p>2</p></td>
<td><p>37507459</p></td>
<td><p>0.08438566</p></td>
<td><p>6.318294e-12</p></td>
<td><p>0.009006862</p></td>
<td><p>-</p></td>
<td><p>C=0.244312/8354 <a href="https://www.ncbi.nlm.nih.gov/snp/rs4670713#frequency_tab">ALFA</a><br />
C=0.171667/103 NorthernSweden<br />
C=0.193825/747 ALSPAC</p></td>
<td><p>0.244312</p></td>
<td><p>0.171667</p></td>
</tr>
<tr class="even">
<td><p>rs6850796</p></td>
<td><p>4</p></td>
<td><p>102624773</p></td>
<td><p>0.06652043</p></td>
<td><p>6.645545e-12</p></td>
<td><p>0.011376648</p></td>
<td><p>-</p></td>
<td><p>C=0.097767/2390 <a href="https://www.ncbi.nlm.nih.gov/snp/rs6850796#frequency_tab">ALFA</a><br />
C=0.004687/21 Estonian<br />
C=0.006012/6 GoNL</p></td>
<td><p>0.097767</p></td>
<td><p>0.004687</p></td>
</tr>
<tr class="odd">
<td><p>rs6967342</p></td>
<td><p>7</p></td>
<td><p>12318735</p></td>
<td><p>0.08582178</p></td>
<td><p>6.949173e-12</p></td>
<td><p>0.011536629</p></td>
<td><p>-</p></td>
<td><p>G=0.029891/569 <a href="https://www.ncbi.nlm.nih.gov/snp/rs6967342#frequency_tab">ALFA</a><br />
G=0.011607/52 Estonian<br />
G=0.011667/7 NorthernSweden</p></td>
<td><p>0.029891</p></td>
<td><p>0.011607</p></td>
</tr>
<tr class="even">
<td><p>rs7957163</p></td>
<td><p>12</p></td>
<td><p>2675893</p></td>
<td><p>0.06637853</p></td>
<td><p>1.019880e-11</p></td>
<td><p>0.011317492</p></td>
<td><p>-</p></td>
<td><p>A=0.071777/10732 <a href="https://www.ncbi.nlm.nih.gov/snp/rs7957163#frequency_tab">ALFA</a><br />
A=0./0 PRJEB36033<br />
A=0.025/1 GENOME_DK</p></td>
<td><p>0.071777</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>

Cluster 2: [ 8 Significant SNPs]

<table>
<thead>
<tr class="header">
<th><p>SNP</p></th>
<th><p>CHR</p></th>
<th><p>BP</p></th>
<th><p>SE</p></th>
<th><p>P-Val</p></th>
<th><p>Beta</p></th>
<th><p>Gene</p></th>
<th><p>Freq</p></th>
<th><p>Max Freq</p></th>
<th><p>Min Freq</p></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><p>rs11731396</p></td>
<td><p>4</p></td>
<td><p>84352339</p></td>
<td><p>0.66790073</p></td>
<td><p>6.22809E-05</p></td>
<td><p>2.674273254</p></td>
<td><p>-</p></td>
<td><p>C=0.144393/29546 <a href="https://www.ncbi.nlm.nih.gov/snp/rs11731396#frequency_tab">ALFA</a><br />
C=0./0 PRJEB36033<br />
C=0.066424/1113 TOMMO</p></td>
<td><p>0.144393</p></td>
<td><p>0</p></td>
</tr>
<tr class="even">
<td><p>rs1595361</p></td>
<td><p>18</p></td>
<td><p>4569102</p></td>
<td><p>0.807969804</p></td>
<td><p>7.40472E-05</p></td>
<td><p>3.201897261</p></td>
<td><p>-</p></td>
<td><p>C=0.425201/10903 <a href="https://www.ncbi.nlm.nih.gov/snp/rs1595361#frequency_tab">ALFA</a><br />
T=0.266667/104 SGDP_PRJ<br />
C=0.266667/8 PRJEB36033</p></td>
<td><p>0.425201</p></td>
<td><p>0.266667</p></td>
</tr>
<tr class="odd">
<td><p>rs3787397</p></td>
<td><p>20</p></td>
<td><p>52700084</p></td>
<td><p>0.645998874</p></td>
<td><p>4.99397E-05</p></td>
<td><p>2.62011272</p></td>
<td><p>DOK5 </p></td>
<td><p>A=0.33283/51972 <a href="https://www.ncbi.nlm.nih.gov/snp/rs3787397#frequency_tab">ALFA</a><br />
A=0.255591/160 Chileans<br />
G=0.261905/11 Siberian</p></td>
<td><p>0.33283</p></td>
<td><p>0.255591</p></td>
</tr>
<tr class="even">
<td><p>rs4670713</p></td>
<td><p>2</p></td>
<td><p>37507459</p></td>
<td><p>0.635650094</p></td>
<td><p>4.44686E-05</p></td>
<td><p>2.595324143</p></td>
<td><p>-</p></td>
<td><p>C=0.244312/8354 <a href="https://www.ncbi.nlm.nih.gov/snp/rs4670713#frequency_tab">ALFA</a><br />
C=0.171667/103 NorthernSweden<br />
C=0.193825/747 ALSPAC</p></td>
<td><p>0.244312</p></td>
<td><p>0.171667</p></td>
</tr>
<tr class="odd">
<td><p>rs5956063</p></td>
<td><p>23</p></td>
<td><p>117486408</p></td>
<td><p>0.509138878</p></td>
<td><p>6.99535E-05</p></td>
<td><p>2.024562099</p></td>
<td><p>-</p></td>
<td><p>T=0.042562/618 <a href="https://www.ncbi.nlm.nih.gov/snp/rs5956063#frequency_tab">ALFA</a><br />
T=0.086535/250 ALSPAC<br />
T=0.087648/325 TWINSUK</p></td>
<td><p>0.042562</p></td>
<td><p>0.086535</p></td>
</tr>
<tr class="even">
<td><p>rs6440278</p></td>
<td><p>3</p></td>
<td><p>146124983</p></td>
<td><p>0.625697778</p></td>
<td><p>7.00049E-05</p></td>
<td><p>2.487942842</p></td>
<td><p>-</p></td>
<td><p>G=0.166145/35014 <a href="https://www.ncbi.nlm.nih.gov/snp/rs6440278#frequency_tab">ALFA</a><br />
G=0.110229/125 Daghestan<br />
G=0.111821/70 Chileans</p></td>
<td><p>0.166145</p></td>
<td><p>0.110229</p></td>
</tr>
<tr class="odd">
<td><p>rs7318474</p></td>
<td><p>13</p></td>
<td><p>28027431</p></td>
<td><p>0.632986484</p></td>
<td><p>4.30473E-05</p></td>
<td><p>2.58922144</p></td>
<td><p>-</p></td>
<td><p>G=0.358709/110575 <a href="https://www.ncbi.nlm.nih.gov/snp/rs7318474#frequency_tab">ALFA</a><br />
G=0.083905/1406 TOMMO<br />
G=0.113208/24 Vietnamese</p></td>
<td><p>0.358709</p></td>
<td><p>0.083905</p></td>
</tr>
<tr class="even">
<td><p>rs7894606</p></td>
<td><p>10</p></td>
<td><p>92956616</p></td>
<td><p>0.758332586</p></td>
<td><p>9.56854E-05</p></td>
<td><p>2.958467684</p></td>
<td><p>PCGF5 </p></td>
<td><p>A=0.17591/5468 <a href="https://www.ncbi.nlm.nih.gov/snp/rs7894606#frequency_tab">ALFA</a><br />
A=0.028037/6 Vietnamese<br />
A=0.048123/141 KOREAN<br />
A=0.061135/112 Korea1K</p></td>
<td><p>0.17591</p></td>
<td><p>0.028037</p></td>
</tr>
</tbody>
</table>

#### Comment on the significant SNPs identified in relation to the PCs and clusters. Were the SNPs different for each PC and for each cluster? Did the analysis using PCs lead to different SNPs from those used with clustering? What genes contain those SNPs (or at least the 10 top significant SNPs)?

For PC1:

-   The top 10 significant SNPs include rs12613530, rs213037, rs235679, rs3181360, rs359428, rs3787397, rs4670713, rs6850796, rs6967342, and rs7957163.

-   Each SNP has associated information such as CHR, BP position, beta coefficient (Beta), p-value (P-Val), standard error (SE), gene, and frequency information.

-   The genes associated with these SNPs include CAPN14, ECE1, NKAIN2, TNFSF8, DOK5, and others.

For Cluster 2:

-   The significant SNPs include rs11731396, rs1595361, rs3787397 (appeared in PC1), rs4670713 (also in PC1), rs5956063, rs6440278, rs7318474, and rs7894606.

-   Similar to the PC1 data, each SNP has associated information including CHR, BP, SE, P-Val, Beta, gene, and frequency information.

-   The genes associated with there SNPs include DOK5 and PCGF5.

Comments:

-   The SNPs identified for each PC and cluster seem to overlap to some extent. For example, rs3787397 and rs4670713 appear in both PC1 and Cluster 2.

-   However, there are also unique SNPs for each group. For example, rs12613530, rs213037, rs235679, rs3181360, rs6850796, rs6967342, and rs7957163 are unique to PC1, while rs11731396, rs1595361, rs5956063, rs6440278, rs7318474, and rs7894606 are unique to Cluster 2.

-   The analysis using PCs and clustering lead to different sets of significant SNPs due to the different statistical methods and grouping criteria used.

-   The genes containing these SNPs include CAPN14, ECE1, NKAIN2, TNFSF8, DOK5, and PCGF5 among others. These genes may have potential biological relevance to the traits or phenotypes being studied.
